core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+                QueryId queryId,
+                PlanNodeId partitionedSourceNodeId,
+                TableExecuteContextManager tableExecuteContextManager,
+                SplitSource splitSource,
+                Multimap<PlanNodeId, ExchangeSourceHandle> replicatedExchangeSourceHandles,
+                int splitBatchSize,
+                LongConsumer getSplitTimeRecorder,
+                Optional<CatalogName> catalogRequirement,
+                int targetPartitionSplitCount)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.partitionedSourceNodeId = requireNonNull(partitionedSourceNodeId, "partitionedSourceNodeId is null");
+            this.tableExecuteContextManager = requireNonNull(tableExecuteContextManager, "tableExecuteContextManager is null");
+            this.splitSource = requireNonNull(splitSource, "splitSource is null");
+            this.replicatedExchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(replicatedExchangeSourceHandles, "replicatedExchangeSourceHandles is null"));
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+            this.catalogRequirement = requireNonNull(catalogRequirement, "catalogRequirement is null");
+            checkArgument(targetPartitionSplitCount > 0, "targetPartitionSplitCount must be positive: %s", targetPartitionSplitCount);
+            this.targetPartitionSplitCount = targetPartitionSplitCount;
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            if (finished || closed) {
+                return ImmutableList.of();
+            }
+
+            while (true) {
+                if (remotelyAccessibleSplitBuffer.size() >= targetPartitionSplitCount) {
+                    ImmutableList.Builder<Split> splits = ImmutableList.builder();
+                    for (int i = 0; i < targetPartitionSplitCount; i++) {
+                        splits.add(remotelyAccessibleSplitBuffer.poll());
+                    }
+                    return ImmutableList.of(
+                            new TaskDescriptor(
+                                    currentPartitionId++,
+                                    ImmutableListMultimap.<PlanNodeId, Split>builder().putAll(partitionedSourceNodeId, splits.build()).build(),
+                                    replicatedExchangeSourceHandles,
+                                    new NodeRequirements(catalogRequirement, Optional.empty())));
+                }
+                for (HostAddress remoteHost : locallyAccessibleSplitBuffer.keySet()) {
+                    Set<Split> hostSplits = locallyAccessibleSplitBuffer.get(remoteHost);
+                    if (hostSplits.size() >= targetPartitionSplitCount) {
+                        List<Split> splits = removeN(hostSplits, targetPartitionSplitCount);
```

Comment:
we should probably take into account `Split.getSplitWeight`. Not necessarily now, but please add a TODO and create a github issue.

Reply:
There's a TODO item in the Github issue: `Implement weight based split assignment`